### PR TITLE
Return new State object on `patch` if server respond with 200

### PR DESCRIPTION
**Pull request by madsmadsen**
Return new State object on `patch` if server respond with 200

### CODE REVIEW

File: src/resource.ts [hunk 1]
```diff
@@ -206,13 +206,25 @@ export class Resource<T = any> extends EventEmitter {
...
     );
 
+    if (response.status === 200) {
+      const state = await this.client.getStateForResponse(this.uri, response);
+      this.updateCache(state);
+
+      return state;
+    }
+
+    this.clearCache();
```
Comment (evert) [inline src/resource.ts:227]: This is already handled centrally for all non-safe methods here: 

https://github.com/badgateway/ketting/blob/master/src/client.ts#L256

Comment (madsmadsen) [reply src/resource.ts]: Ahh, didn't notice that.

When is that run before or after this?

**Comment by evert** [comment]
Generally I think this is a good idea, but `PATCH` does not have standard semantics for response bodies. The response body to a `PATCH` does not imply that it's the new resource state. There's for example nothing stopping servers from returning `{success: true}`

However, there is a way for a sever to tell the client that it is, using the `Prefer: return=representation` and `Preference-Applied` headers. This is tracked as a ticket here #22 .

If you're interested in implementing that, even if it's just for `PATCH`, I'd definitely welcome to PR.

**Comment by madsmadsen** [comment]
You're right, did some reading on the `PATCH` RFC and it does indeed not say anything about which response to expect.
Their example does only includes a `204 No content`.

But I think it would make sense to have the `PATCH` request return the response of 200 and 201 responses, as they could include some information a client needs before making a new GET request

**Comment by evert** [comment]
@madsmadsen I do generally agree that it's useful to have access to this, but to automatically assume that the response is a full new state for the resource, we can only do that with the `Preference-Applied` present.

I don't see a problem returning the state though, as long as we don't update the cache. I believe that `post()` does this too

**Comment by madsmadsen** [comment]
> @madsmadsen I do generally agree that it's useful to have access to this, but to automatically assume that the response is a full new state for the resource, we can only do that with the `Preference-Applied` present.
> 
> I don't see a problem returning the state though, as long as we don't update the cache. I believe that `post()` does this too

Makes sense, updated the PR so it just returns the State if the status is 200

**Comment by evert** [comment]
:+1:

**Pull request merged**